Poetry plugin docs

grahamc opened the pull request.

adisbladis committed:
Migrate to poetry2nix

adisbladis committed:
Deployment sorting fails with TypeError when name is None

name is of type Optional[str] and can therefore be None.
In the `nixops info` output we default to the deployment UUID so it
feels reasonable to fall back sorting on UUID.

grahamc committed:
include nix

adisbladis committed:
Drop setup.py

It's been replaced by pyproject.toml

adisbladis committed:
Add black to pyproject.toml

adisbladis committed:
nix_expr: Make list ordering deterministic

adisbladis committed:
Clean up release.nix

adisbladis committed:
Re-add manual build

adisbladis committed:
pyproject.toml: Add lxml for mypy type ratchet

adisbladis committed:
mypy-ratchet: Run first ratchet run inside a nix-shell

We need to evaluate twice, once using any previous version of mypy and
a second time using the current version of mypy.

If we don't a mypy version bump can find new type errors that makes
the initial check fail.

grahamc committed:
Initial plugin authoring docs

grahamc committed:
special note on mypy and black

grahamc committed:
fixup rst

grahamc committed:
wat

grahamc committed:
Fixup

grahamc committed:
fixup

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,221 @@
+Authoring a Plugin
+====
+
+NixOps plugins extend NixOps' core to support additional hosting
```
```suggestion
NixOps plugins extend NixOps core to support additional hosting
```

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,221 @@
+a mypy configuration at least as strict as the NixOps mypy
+configuration.
+
+First, create a ``pyproject.toml`` to describe your project. This is
```
Should we maybe add a reference to https://www.python.org/dev/peps/pep-0517/ ?

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,221 @@
+
+Create a ``default.nix``::
+
+  { nixpkgs ? <nixpkgs>
```
I'd skip `nixpkgs ? <nixpkgs>` and just do `pkgs ? import <nixpkgs> {}`.
The only reason it was done this way in the nixops repo was so I didn't have to change how the man page was built.

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,221 @@
+  let
+    overrides = import ./overrides.nix { inherit pkgs; };
+  in pkgs.poetry2nix.mkPoetryApplication {
+    src = pkgs.lib.cleanSource ./.;
```
I think we should encourage using `projectDir` instead of the separate params.
This will apply the default gitignore/pycache source filters and "does the right thing".

grahamc replied:
I agree, but at the time of writing, the 20.03 branch doesn't support it. As soon as the channel contains it, I'll update the docs

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,221 @@
+  { pkgs }:
+
+  self: super: {
+    nixops = super.nixops.overridePythonAttrs({ nativeBuildInputs ? [], ... }: {
```
Note to self: We should add this to `poetry2nix` upstream.

grahamc committed:
nix-expr.RawValue: add types

Based on:

Grepping the source code for RawValue, I see they are all strings.

grahamc committed:
nix-expr.MultiLineRawValue: add types

Based on grepping the source code for MultiLineRawValue, it apears they're all just a list of RawValue values, ie: str's

grahamc committed:
Container: add some type hints based on other Value objects

grahamc committed:
Update doc/plugins/authoring.rst

Co-Authored-By: adisbladis <adisbladis@gmail.com>

grahamc committed:
Link to PEP-0517

grahamc committed:
default.nix: update pkgs import

grahamc commented:
Updated with the feedback I could address so far :)

grahamc committed:
switch to projectDir

grahamc commented:
Updated to use the 20.03 updates to poetry2nix!

adisbladis commented on doc/plugins/authoring.rst:
```diff
@@ -0,0 +1,211 @@
+    buildInputs = [
+      (pkgs.poetry2nix.mkPoetryEnv {
+        projectDir = ./.;
+        overrides = pkgs.poetry2nix.overrides.withoutDefaults overrides;
```
Why `withoutDefaults` ?

grahamc replied:
I'm not sure, I copy-pasted it from the wrong section of the poetry2nix docs by misake :) Can you fix it?

adisbladis commented:
Squashed and merged.
